src/jitlayers.cpp

```diff
@@ -0,0 +1,396 @@
+            Mangler::getNameWithPrefix(MangledNameStream, Name, DL);
+        }
+        return MangledName;
+    }
+
+    void addGlobalMapping(StringRef Name, void *Addr) {
+       GlobalSymbolTable[mangle(Name)] = Addr;
+    }
+
+    ModuleHandleT addModule(Module *M) {
+        // We need a memory manager to allocate memory and resolve symbols for this
+        // new module. Create one that resolves symbols by looking back into the
+        // JIT.
+        auto Resolver = orc::createLambdaResolver(
+                          [&](const std::string &Name) {
+                            // TODO: consider moving the FunctionMover resolver here
+                            // Step 0: ObjectLinkingLayer has checked whether it is in the current module
+                            // Step 1: Check against list of known external globals
+                            GlobalSymbolTableT::const_iterator pos = GlobalSymbolTable.find(Name);
+                            if (pos != GlobalSymbolTable.end())
+                                return RuntimeDyld::SymbolInfo((intptr_t)pos->second, JITSymbolFlags::Exported);
+                            // Step 2: Search all previously emitted symbols
+                            if (auto Sym = findSymbol(Name))
+                              return RuntimeDyld::SymbolInfo(Sym.getAddress(),
+                                                             Sym.getFlags());
+                            // Step 2: Search the program symbols
+                            if (uint64_t addr = SectionMemoryManager::getSymbolAddressInProcess(Name))
+                                return RuntimeDyld::SymbolInfo(addr, JITSymbolFlags::Exported);
+                            // Return failure code
+                            return RuntimeDyld::SymbolInfo(nullptr);
+                          },
+                          [](const std::string &S) { return nullptr; }
+                        );
+        SmallVector<std::unique_ptr<Module>,1> Ms;
+        Ms.push_back(std::unique_ptr<Module>{M});
+        return CompileLayer->addModuleSet(std::move(Ms),
+                                         MemMgr,
+                                         std::move(Resolver));
+    }
+
+    void removeModule(ModuleHandleT H) { CompileLayer->removeModuleSet(H); }
+
+    orc::JITSymbol findSymbol(const std::string &Name) {
+        return CompileLayer->findSymbol(Name, true);
+    }
+
+    orc::JITSymbol findUnmangledSymbol(const std::string Name) {
+        return findSymbol(mangle(Name));
+    }
+
+    uint64_t getGlobalValueAddress(const std::string &Name) {
+        return CompileLayer->findSymbol(mangle(Name), false).getAddress();
+    }
+
+    uint64_t getFunctionAddress(const std::string &Name) {
+        return CompileLayer->findSymbol(mangle(Name), false).getAddress();
+    }
+
+    uint64_t FindFunctionNamed(const std::string &Name) {
+        return 0; // Functions are not kept around
```

Comment:
Thinking again, I don't think there's actually any case any more where we need to look in any module other than the source module (during bootstrap that will always be the shadow module), so probably just ifdef out this check in ORCJIT mode.